zkvm/src/program.rs

```diff
@@ -0,0 +1,176 @@
+#[derive(Clone, Debug)]
+pub struct Program(Vec<Instruction>);
+
+macro_rules! def_op {
+    ($func_name:ident, $op:ident) => (
+           /// Adds a `$func_name` instruction.
+           pub fn $func_name(&mut self) -> &mut Program{
+             self.0.push(Instruction::$op);
+             self
+        }
+    );
+    ($func_name:ident, $op:ident, $type:ty) => (
+           /// Adds a `$func_name` instruction.
+           pub fn $func_name(&mut self, arg :$type) -> &mut Program{
+             self.0.push(Instruction::$op(arg));
+             self
+        }
+    );
+    ($func_name:ident, $op:ident, $type1:ty, $type2:ty) => (
+           /// Adds a `$func_name` instruction.
+           pub fn $func_name(&mut self, arg1: $type1, arg2: $type2) -> &mut Program{
+             self.0.push(Instruction::$op(arg1, arg2));
+             self
+        }
+    );
+}
+
+impl Program {
+    def_op!(add, Add);
+    def_op!(alloc, Alloc, Option<ScalarWitness>);
+    def_op!(and, And);
+    def_op!(borrow, Borrow);
+    def_op!(call, Call);
+    def_op!(cloak, Cloak, usize, usize);
+    def_op!(r#const, Const);
+    def_op!(contract, Contract, usize);
+    def_op!(delegate, Delegate);
+    def_op!(drop, Drop);
+    def_op!(dup, Dup, usize);
+    def_op!(eq, Eq);
+    def_op!(export, Export);
+    def_op!(expr, Expr);
+    def_op!(import, Import);
+    def_op!(input, Input);
+    def_op!(issue, Issue);
+    def_op!(log, Log);
+    def_op!(maxtime, Maxtime);
+    def_op!(mintime, Mintime);
+    def_op!(mul, Mul);
+    def_op!(neg, Neg);
+    def_op!(nonce, Nonce);
+    def_op!(or, Or);
+    def_op!(output, Output, usize);
+    def_op!(range, Range, BitRange);
+    def_op!(retire, Retire);
+    def_op!(roll, Roll, usize);
+    def_op!(select, Select, u8, u8);
+    def_op!(sign_tx, Signtx);
+    def_op!(unblind, Unblind);
+    def_op!(var, Var);
+    def_op!(verify, Verify);
+
+    /// Creates an empty `Program`.
+    pub fn new() -> Self {
+        Program(vec![])
+    }
+
+    /// Creates an empty `Program` and passes its &mut to the closure to let it add the instructions.
+    /// Returns the resulting program.
+    pub fn build<F>(builder: F) -> Self
+    where
+        F: FnOnce(&mut Self) -> &mut Self,
+    {
+        let mut program = Self::new();
+        builder(&mut program);
+        program
+    }
+
+    /// Creates a program from parsing the opaque data slice of encoded instructions.
+    pub fn parse(data: &[u8]) -> Result<Self, VMError> {
+        SliceReader::parse(data, |r| {
+            let mut program = Self::new();
+            while r.len() > 0 {
+                program.0.push(Instruction::parse(r)?);
+            }
+            Ok(program)
+        })
+    }
+
+    /// Converts the program to a plain vector of instructions.
+    pub fn to_vec(self) -> Vec<Instruction> {
+        self.0
+    }
+
+    /// Returns the serialized length of the program.
+    pub fn serialized_length(&self) -> usize {
```

Comment:
Since we're exposing `program` as the public module, it seems like we do not need to expose `serialized_length` and `encode` as `pub` since a caller interested in building programs shouldn't need to call these functions!

Reply:
same with `parse`?